benchmark/kvs/kvs.go

```diff
@@ -0,0 +1,139 @@
+package kvs
+
+import (
+	"fmt"
+	"github.com/UBC-NSS/pgo/distsys"
+	"github.com/UBC-NSS/pgo/distsys/tla"
+)
+
+var _ = new(fmt.Stringer) // unconditionally prevent go compiler from reporting unused fmt import
+var _ = distsys.ErrDone
+var _ = tla.TLAValue{} // same, for tla
+
+func Key1(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.MakeTLAString("key1")
+}
+func Key2(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.MakeTLAString("key2")
+}
+func KeySet(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.MakeTLASet(Key1(iface), Key2(iface))
+}
+func Value1(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.MakeTLAString("value1")
+}
+func Value2(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.MakeTLAString("value2")
+}
+func ValueSet(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.MakeTLASet(Value1(iface), Value2(iface))
+}
+func WriteNodeSet(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.TLA_DotDotSymbol(tla.MakeTLANumber(1), iface.GetConstant("NumWriteNodes")())
+}
+func ReadNodeSet(iface distsys.ArchetypeInterface) tla.TLAValue {
+	return tla.TLA_DotDotSymbol(tla.TLA_PlusSymbol(iface.GetConstant("NumWriteNodes")(), tla.MakeTLANumber(1)), tla.TLA_PlusSymbol(iface.GetConstant("NumWriteNodes")(), iface.GetConstant("NumReadNodes")()))
+}
+
+var procTable = distsys.MakeMPCalProcTable()
+
+var jumpTable = distsys.MakeMPCalJumpTable(
+	distsys.MPCalCriticalSection{
+		Name: "AWriteNode.writeNodeLoop",
+		Body: func(iface distsys.ArchetypeInterface) error {
+			var err error
+			_ = err
+			kvs, err := iface.RequireArchetypeResourceRef("AWriteNode.kvs")
+			if err != nil {
+				return err
+			}
+			if tla.TLA_TRUE.AsBool() {
+				var keyRead = KeySet(iface)
+				if keyRead.AsSet().Len() == 0 {
+					return distsys.ErrCriticalSectionAborted
+				}
+				var key tla.TLAValue = keyRead.SelectElement(iface.NextFairnessCounter("AWriteNode.writeNodeLoop.0", uint(keyRead.AsSet().Len())))
+				_ = key
+				var valRead = ValueSet(iface)
+				if valRead.AsSet().Len() == 0 {
+					return distsys.ErrCriticalSectionAborted
+				}
+				var val tla.TLAValue = valRead.SelectElement(iface.NextFairnessCounter("AWriteNode.writeNodeLoop.1", uint(valRead.AsSet().Len())))
+				_ = val
+				err = iface.Write(kvs, []tla.TLAValue{key}, val)
+				if err != nil {
+					return err
+				}
+				return iface.Goto("AWriteNode.writeNodeLoop")
```

Comment:
@fhackett 
It looks like this code was generated from:
```
with (key \in KeySet, val \in ValueSet) {
   kvs[key] := val;
};
```
However this seems to have different semantics in go and pcal:
- go iterates over pairs of the "zipped" lists (keys, values)
- pcal iterates over pairs of the cartesian product (i.e similar to how this would look in Scala)

Maybe this is a bug?

Reply:
Took me a bit to figure out (thanks Shayan drawing my attention to this!), and yes it is a bug.

It actually means PGo's "branch fairness" system as a whole is flawed, not specifically this case of it...

Noted in #207.